File: src/types/host.ts
import type { EpochMillis, GroupId, HostId, TagId } from "./common";

export type AuthType = "password" | "private_key";

export type ConnectionStatus =
  | "connected"
  | "disconnected"
  | "failed"
  | "auth_failed"
  | "timeout"
  | "network_error";

export type TagDto = {
  id: TagId;
  name: string;
  color?: string;
  createdAt: EpochMillis;
  updatedAt: EpochMillis;
};

export type HostGroupDto = {
  id: GroupId;
  name: string;
  parentId?: GroupId;
  sortOrder: number;
  createdAt: EpochMillis;
  updatedAt: EpochMillis;
};

export type HostDto = {
  id: HostId;
  name: string;
  hostname: string;
  port: number;
  username: string;
  authType: AuthType;
  hasPassword: boolean;
  privateKeyPath?: string;
  hasPassphrase: boolean;
  groupId?: GroupId;
  tags: TagDto[];
  connectTimeoutMs: number;
  keepaliveIntervalSecs: number;
  startupCommand?: string;
  terminalTheme?: string;
  lastConnectedAt?: EpochMillis;
  lastStatus?: ConnectionStatus;
  lastErrorMessage?: string;
  createdAt: EpochMillis;
  updatedAt: EpochMillis;
};

export type SecretUpdate =
  | { action: "keep" }
  | { action: "replace"; value: string }
  | { action: "clear" };

export type CreateHostInput = {
  name: string;
  hostname: string;
  port: number;
  username: string;
  authType: AuthType;
  password?: string;
  privateKeyPath?: string;
  privateKeyPassphrase?: string;
  groupId?: GroupId;
  tagIds?: TagId[];
  connectTimeoutMs?: number;
  keepaliveIntervalSecs?: number;
  startupCommand?: string;
  terminalTheme?: string;
};

export type UpdateHostInput = {
  name?: string;
  hostname?: string;
  port?: number;
  username?: string;
  authType?: AuthType;
  password?: SecretUpdate;
  privateKeyPath?: string;
  privateKeyPassphrase?: SecretUpdate;
  groupId?: GroupId | null;
  tagIds?: TagId[];
  connectTimeoutMs?: number;
  keepaliveIntervalSecs?: number;
  startupCommand?: string | null;
  terminalTheme?: string | null;
};

export type DuplicateHostInput = {
  hostId: HostId;
  name: string;
};

export type HostFilter = {
  keyword?: string;
  groupId?: GroupId;
  tagIds?: TagId[];
  authType?: AuthType;
  recentlyConnected?: boolean;
};

export type HostListResult = {
  hosts: HostDto[];
};

File: src-tauri/src/contracts/host.rs
use crate::contracts::{EpochMillis, GroupId, HostId, TagId};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AuthType {
    Password,
    PrivateKey,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionStatus {
    Connected,
    Disconnected,
    Failed,
    AuthFailed,
    Timeout,
    NetworkError,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TagDto {
    pub id: TagId,
    pub name: String,
    pub color: Option<String>,
    pub created_at: EpochMillis,
    pub updated_at: EpochMillis,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HostDto {
    pub id: HostId,
    pub name: String,
    pub hostname: String,
    pub port: u16,
    pub username: String,
    pub auth_type: AuthType,
    pub has_password: bool,
    pub private_key_path: Option<String>,
    pub has_passphrase: bool,
    pub group_id: Option<GroupId>,
    pub tags: Vec<TagDto>,
    pub connect_timeout_ms: u64,
    pub keepalive_interval_secs: u64,
    pub startup_command: Option<String>,
    pub terminal_theme: Option<String>,
    pub last_connected_at: Option<EpochMillis>,
    pub last_status: Option<ConnectionStatus>,
    pub last_error_message: Option<String>,
    pub created_at: EpochMillis,
    pub updated_at: EpochMillis,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateHostInput {
    pub name: String,
    pub hostname: String,
    pub port: u16,
    pub username: String,
    pub auth_type: AuthType,
    pub password: Option<String>,
    pub private_key_path: Option<String>,
    pub private_key_passphrase: Option<String>,
    pub group_id: Option<GroupId>,
    pub tag_ids: Option<Vec<TagId>>,
    pub connect_timeout_ms: Option<u64>,
    pub keepalive_interval_secs: Option<u64>,
    pub startup_command: Option<String>,
    pub terminal_theme: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateHostInput {
    pub name: Option<String>,
    pub hostname: Option<String>,
    pub port: Option<u16>,
    pub username: Option<String>,
    pub auth_type: Option<AuthType>,
    pub password: Option<SecretUpdate>,
    pub private_key_path: Option<String>,
    pub private_key_passphrase: Option<SecretUpdate>,
    pub group_id: Option<Option<GroupId>>,
    pub tag_ids: Option<Vec<TagId>>,
    pub connect_timeout_ms: Option<u64>,
    pub keepalive_interval_secs: Option<u64>,
    pub startup_command: Option<Option<String>>,
    pub terminal_theme: Option<Option<String>>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DuplicateHostInput {
    pub host_id: HostId,
    pub name: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "snake_case", tag = "action", content = "value")]
pub enum SecretUpdate {
    Keep,
    Replace(String),
    Clear,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct HostFilter {
    pub keyword: Option<String>,
    pub group_id: Option<GroupId>,
    pub tag_ids: Option<Vec<TagId>>,
    pub auth_type: Option<AuthType>,
    pub recently_connected: Option<bool>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct HostListResult {
    pub hosts: Vec<HostDto>,
}
